src/core/server/logging/appenders/rolling_file/rolling_file_appender.ts

```diff
@@ -0,0 +1,174 @@
+   * The {@link TriggeringPolicy | policy} to use to determine if a rollover should occur.
+   */
+  policy: TriggeringPolicyConfig;
+  /**
+   * The {@link RollingStrategy | rollout strategy} to use for rolling.
+   */
+  strategy: RollingStrategyConfig;
+}
+
+/**
+ * Appender that formats all the `LogRecord` instances it receives and writes them to the specified file.
+ * @internal
+ */
+export class RollingFileAppender implements DisposableAppender {
+  public static configSchema = schema.object({
+    kind: schema.literal('rolling-file'),
+    layout: Layouts.configSchema,
+    path: schema.string(),
+    policy: triggeringPolicyConfigSchema,
+    strategy: rollingStrategyConfigSchema,
+  });
+
+  private isRolling = false;
+  private disposed = false;
+  private rollingPromise?: Promise<void>;
+
+  private readonly layout: Layout;
+  private readonly context: RollingFileContext;
+  private readonly fileManager: RollingFileManager;
+  private readonly policy: TriggeringPolicy;
+  private readonly strategy: RollingStrategy;
+  private readonly buffer: BufferAppender;
+
+  constructor(config: RollingFileAppenderConfig) {
+    this.context = new RollingFileContext(config.path);
+    this.context.refreshFileInfo();
+    this.fileManager = new RollingFileManager(this.context);
+    this.layout = Layouts.create(config.layout);
+    this.policy = createTriggeringPolicy(config.policy, this.context);
+    this.strategy = createRollingStrategy(config.strategy, this.context);
+    this.buffer = new BufferAppender();
+  }
+
+  /**
+   * Formats specified `record` and writes it to the specified file. If the record
+   * would trigger a rollover, it will be performed before the effective write operation.
+   */
+  public append(record: LogRecord) {
+    // if we are currently rolling the files, push the log record
+    // into the buffer, which will be flushed once rolling is complete
+    if (this.isRolling) {
+      this.buffer.append(record);
+      return;
+    }
+    if (this.needRollout(record)) {
+      this.buffer.append(record);
+      this.rollingPromise = this.performRollout();
+      return;
+    }
+
+    this._writeToFile(record);
+  }
+
+  private _writeToFile(record: LogRecord) {
+    this.fileManager.write(`${this.layout.format(record)}\n`);
+  }
+
+  /**
+   * Disposes the appender.
+   * If a rollout is currently in progress, it will be awaited.
+   */
+  public async dispose() {
+    if (this.disposed) {
+      return;
+    }
+    this.disposed = true;
+    if (this.rollingPromise) {
+      await this.rollingPromise;
+    }
+    await this.buffer.dispose();
+    await this.fileManager.closeStream();
+  }
+
+  private async performRollout() {
+    if (this.isRolling) {
+      return;
+    }
+    this.isRolling = true;
+    try {
+      await this.strategy.rollout();
+      await this.fileManager.closeStream();
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.log('Error while rolling file: ', e);
+    }
```

Comment:
> Would a failed rollout never, sometimes, or always prevent future logs from being written?

it's never `never` when you interact with filesystem, but except in case of hard I/O failure, I think the worth case would be that we would not be able to roll the log file, and that the next rolling cycle would still write in the previous (non-renamed) file. Which seemed acceptable (also we can't really do anything except warning about the failure)